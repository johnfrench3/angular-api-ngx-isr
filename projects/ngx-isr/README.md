# Incremental Static Regeneration for Angular
Incremental Static Regeneration (ISR) enables developers and content editors to use static-generation on a per-page basis, without needing to rebuild the entire site. With ISR, you can retain the benefits of static while scaling to millions of pages. [Source](https://vercel.com/docs/concepts/next.js/incremental-static-regeneration)

ðŸ“° [Read the blog post](https://itnext.io/incremental-static-regeneration-for-angular-42b0a8440e53)

# How to use it?

1. Install npm package
```bash
npm install ngx-isr
```

> Make sure you have Angular Universal in your app.

> Setup Angular Universal: `ng add @nguniversal/express-engine`


2. Initialize `ISRHandler` inside `server.ts`
```ts
const isr = new ISRHandler({
  indexHtml,
  invalidateSecretToken: 'MY_TOKEN', // replace with env secret key
  enableLogging: !environment.production
});
```

3. Add invalidation url handler
```ts
server.get("/api/invalidate", async (req, res) => await isr.invalidate(req, res));
```

4. Replace Angular default server side rendering with ISR rendering

Replace
```ts
server.get('*',
  (req, res) => {
    res.render(indexHtml, { req, providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }] });
  }
);
```
with
```ts
server.get('*',
  // Serve page if it exists in cache
  async (req, res, next) => await isr.serveFromCache(req, res, next),
  // Server side render the page and add to cache if needed
  async (req, res, next) => await isr.render(req, res, next),
);
```

You can also pass `providers` to each of the `ISRHandler` methods.

```ts
server.get('*',
  ...
  async (req, res, next) => await isr.render(req, res, next, {
    providers: [
      { provide: APP_BASE_HREF, useValue: req.baseUrl }, // <-- Needs to be provided when passing providers 
      { provide: CUSTOM_TOKEN, useValue: 'Hello from ISR' },
      CustomService
    ]
  }),
);
```

ISRHandler provides `APP_BASE_HREF` by default. And if you want pass `providers` into the methods of ISRHandler, you will also have to provide `APP_BASE_HREF` token.

5. Add `revalidate` key in route data

Example:
```ts
{
  path: "example",
  component: ExampleComponent,
  data: { revalidate: 5 },
}
```

> **NOTE:** Routes that don't have revalidate key in data won't be handled by ISR. They will fallback to Angular default server side rendering pipeline.


6. Add `NgxIsrModule` in AppServerModule imports
```ts
import { NgxIsrModule } from 'ngx-isr'; // <-- Import module from library

@NgModule({
  imports: [
    ...
    NgxIsrModule.forRoot()  // <-- Use it in module imports
  ]
})
export class AppServerModule {}
```

When importing the module, `NgxIsrService` will be initialized and will start to listen to route changes, only on the server side, so the browser bundle won't contain any extra code.

# Play with demo

```bash
git clone https://github.com/eneajaho/ngx-isr.git
npm install
npm run dev:ssr
```

1. **Disable Javascript** (in order to not load Angular after the page loads)
2. Navigate to different pages and check how the cache gets regenerated in terminal logs


# How it works?
- [Explanation video](https://vimeo.com/687530247) (w/o voice)

The first time a user opens a pages, we server-side render that page, and save its result in cache.

Next time when a user requests the same page he will be served the first cached response.

To handle Incremental Static Regeneration, we need to configure it from our route data.

For example:
```ts
const routes: Routes = [
  {
    path: "one",
    component: PageOneComponent,
  },
  {
    path: "two",
    component: PageTwoComponent,
    data: { revalidate: 5 },
  },
  {
    path: "three",
    component: PageThreeComponent,
    data: { revalidate: 0 },
  },
];
```

- Path `/one` won't be cached at all, and everytime it is requested it will be server-rendered and then will be served to the user.

- Path `/two` on the first request will be server-rendered and then will be cached. On the second request to it, the user will be served the cache that we got on the first request.
  The url will be added to a regeneration queue, in order to re-generate the cache after `5` seconds.
  On the third request to the same url, if the regeneration was finished the user will be served the regenerated page otherwise he will be served with the old cached page.

- Path `/three` after the first request that is server-rendered, the page will be added to cache and the cache will never be deleted automatically as in path `/two`. So after the first request, all the other ones will come from the cache.


## Changelog
- Version 0.2.0
  * Features:
    * Added `skipCachingOnHttpError` option. It will be enabled by default.
  * Breaking changes:
    * When adding `NgxIsrModule` in AppServerModule imports, we should change it to be `NgxIsrModule.forRoot()`.

## License
MIT
