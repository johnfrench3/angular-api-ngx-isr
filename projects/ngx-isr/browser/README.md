# ngx-isr/browser

Secondary entry point of `ngx-isr`. It can be used by importing from `ngx-isr/browser`.
