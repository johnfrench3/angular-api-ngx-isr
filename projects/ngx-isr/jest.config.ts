/* eslint-disable */
export default {
  displayName: 'ngx-isr',
  preset: '../../jest.preset.js',
  coverageDirectory: '../../coverage/projects/ngx-isr',
};
