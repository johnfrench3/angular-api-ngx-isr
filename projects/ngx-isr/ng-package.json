{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/ngx-isr",
  "lib": {
    "entryFile": "src/public-api.ts"
  },
  "allowedNonPeerDependencies": ["ng-morph"]
}
