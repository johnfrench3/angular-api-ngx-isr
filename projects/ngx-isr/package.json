{
  "name": "ngx-isr",
  "author": "Enea Jahollari",
  "description": "Incremental Static Regeneration for Angular",
  "version": "0.6.0",
  "engines": {
    "node": "^16.13.0 || ^18.10.0",
    "npm": "^9.5.0",
    "yarn": "^1.22.0"
  },
  "peerDependencies": {
    "@angular/common": ">=15.0.0 || >=16.0.0",
    "@angular/core": ">=15.0.0 || >=16.0.0",
    "@types/node": "^14.15.0 || ^16.0.0 || ^18.0.0"
  },
  "dependencies": {
    "tslib": "^2.3.0",
    "ng-morph": "^2.1.3"
  },
  "keywords": [
    "Angular",
    "Library",
    "TypeScript",
    "Angular Universal",
    "Incremental Static Regeneration",
    "Angular ISR"
  ],
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/eneajaho/ngx-isr"
  },
  "schematics": "./schematics/collection.json",
  "ng-add": {
    "save": "dependencies"
  },
  "ng-update": {
    "migrations": "./schematics/migration.json"
  }
}
