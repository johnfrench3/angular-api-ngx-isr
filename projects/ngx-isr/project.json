{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "library",
  "sourceRoot": "projects/ngx-isr/src",
  "prefix": "lib",
  "targets": {
    "build-package": {
      "executor": "@angular-devkit/build-angular:ng-packagr",
      "options": {
        "project": "projects/ngx-isr/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "projects/ngx-isr/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "projects/ngx-isr/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production",
      "outputs": ["{workspaceRoot}/dist/ngx-isr"]
    },
    "build": {
      "executor": "nx:run-commands",
      "options": {
        "parallel": false,
        "commands": [
          {"command": "nx run ngx-isr:build-package"},
          {"command": "tsc -p ./projects/ngx-isr/tsconfig.schematics.json"},
          {"command": "copyfiles projects/ngx-isr/schematics/collection.json dist/ngx-isr/schematics -f"},
          {"command": "copyfiles projects/ngx-isr/schematics/migration.json dist/ngx-isr/schematics -f"}
        ]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "projects/ngx-isr/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
