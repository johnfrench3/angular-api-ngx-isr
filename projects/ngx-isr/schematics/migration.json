{
  "$schema": "../../../node_modules/@angular-devkit/schematics/collection-schema.json",
  "schematics": {
    "migrate-to-rx-angular-isr": {
      "version": "0.6.0",
      "description": "Migrate to @rx-angular/isr and new class names",
      "factory": "./src/migrations/migrate-to-rx-angular-isr/index"
    }
  }
}
