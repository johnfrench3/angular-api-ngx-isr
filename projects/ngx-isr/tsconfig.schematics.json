{
  "compilerOptions": {
    "baseUrl": ".",
    "lib": ["ES2020"],
    "declaration": true,
    "module": "commonjs",
    "moduleResolution": "node",
    "noEmitOnError": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "noUnusedParameters": true,
    "noUnusedLocals": true,
    "rootDir": "schematics",
    "outDir": "../../dist/ngx-isr/schematics",
    "skipDefaultLibCheck": true,
    "skipLibCheck": true,
    "sourceMap": false,
    "strictNullChecks": true,
    "target": "es6",
    "types": ["node"]
  },
  "include": ["schematics/**/*.ts", "schematics/**/*.json"],
  "exclude": ["schematics/**/*.spec.ts", "schematics/**/*.test.ts"]
}
