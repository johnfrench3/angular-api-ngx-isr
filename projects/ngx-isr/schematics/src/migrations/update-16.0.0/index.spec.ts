import {Tree} from '@angular-devkit/schematics';
import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as path from 'path';

describe('isr migration update-16.0.0', () => {
  let appTree: UnitTestTree | undefined;

  it('should replace ngx-isr/server', async () => {
    appTree = await setupTestFile(`
      import { BrowserModule } from '@angular/platform-browser';
      import { NgxIsrModule,
               provideISR,
               ISRHandler,
               FileSystemCacheHandler,
               FileSystemCacheOptions
             } from 'ngx-isr/server';
      import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent,
        ],
        imports: [
          BrowserModule
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
  `);

    const file = appTree?.readContent('app.module.ts');

    // @ts-ignore
    expect(file).toMatchSnapshot();
  });

  it('should replace ngx-isr/models', async () => {
    appTree = await setupTestFile(`
      import { BrowserModule } from '@angular/platform-browser';
      import { NgxIsrService } from 'ngx-isr/browser';
      import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent,
        ],
        imports: [
          BrowserModule
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
  `);

    const file = appTree?.readContent('app.module.ts');

    // @ts-ignore
    expect(file).toMatchSnapshot();
  });

  it('should replace ngx-isr/browser', async () => {
    appTree = await setupTestFile(`
      import { BrowserModule } from '@angular/platform-browser';
      import { NgxIsrService } from 'ngx-isr/browser';
      import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent,
        ],
        imports: [
          BrowserModule
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
  `);

    const file = appTree?.readContent('app.module.ts');

    // @ts-ignore
    expect(file).toMatchSnapshot();
  });

  function setupTestFile(fileInput: string, filePath = './app.module.ts') {
    const runner = new SchematicTestRunner(
      'ngx-isr/server',
      path.join(__dirname, '../../../migration.json')
    );
    const tree = new UnitTestTree(Tree.empty());

    tree.create(filePath, fileInput);

    return runner.runSchematic(`update-16.0.0`, {}, tree);
  }
});
