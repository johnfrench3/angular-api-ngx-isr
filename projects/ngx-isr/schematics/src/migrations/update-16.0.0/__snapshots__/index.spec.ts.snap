// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isr migration update-16.0.0 should add new and remove old dependencies 1`] = `
"{
  "dependencies": {
    "@rx-angular/isr": "16.0.0"
  }
}
"
`;

exports[`isr migration update-16.0.0 should replace ngx-isr/browser 1`] = `
"import { BrowserModule } from '@angular/platform-browser';
import { NgxIsrService } from 'ngx-isr/browser';
import { AppComponent } from './app.component';

@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule],
  bootstrap: [AppComponent],
})
export class AppModule {}
"
`;

exports[`isr migration update-16.0.0 should replace ngx-isr/models 1`] = `
"import { BrowserModule } from '@angular/platform-browser';
import { NgxIsrService } from 'ngx-isr/browser';
import { AppComponent } from './app.component';

@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule],
  bootstrap: [AppComponent],
})
export class AppModule {}
"
`;

exports[`isr migration update-16.0.0 should replace ngx-isr/server 1`] = `
"import {
  NgxIsrModule,
  provideISR,
  ISRHandler,
  FileSystemCacheHandler,
  FileSystemCacheOptions,
} from '@rx-angular/isr/server';

import { BrowserModule } from '@angular/platform-browser';
import { AppComponent } from './app.component';

@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule],
  bootstrap: [AppComponent],
})
export class AppModule {}
"
`;
